internal/daemon/worker/status.go

```diff
@@ -191,14 +194,29 @@ func (w *Worker) sendWorkerStatus(cancelCtx context.Context) {
 	} else {
 		w.updateTags.Store(false)
 		// This may be nil if we are in a multiple hop scenario
+		var addrs []resolver.Address
 		if len(result.CalculatedUpstreams) > 0 {
-			addrs := make([]resolver.Address, 0, len(result.CalculatedUpstreams))
+			addrs = make([]resolver.Address, 0, len(result.CalculatedUpstreams))
 			for _, v := range result.CalculatedUpstreams {
 				addrs = append(addrs, resolver.Address{Addr: v.Address})
 			}
-			w.Resolver().UpdateState(resolver.State{Addresses: addrs})
+			lastStatus := w.lastStatusSuccess.Load().(*LastStatusInformation)
+			// Compare upstreams; update resolver if there is a difference, and emit an event with old and new addresses
+			if lastStatus != nil && upstreamsHasChanged(lastStatus.LastCalculatedUpstreams, addrs) {
```

Comment:
If using []string as what's being stored you can get rid of a lot of the logic here and all of `upstreamsHasChanged` by using https://github.com/hashicorp/go-secure-stdlib/blob/main/strutil/strutil.go#L295